docs/couler-api-design.md

```diff
@@ -0,0 +1,87 @@
+# Couler API Design
+
+This document outlines the design of core Couler APIs to support multiple workflow backends.
+
+## Goals
+
+* Design the core Couler APIs that can be implemented across different workflow engines.
+* Provide a *minimal* set of APIs that are forward-looking, engine-agnostic, and less likely to deprecate over-time.
+
+## Non-Goals
+
+* Provide an exhaustive set of APIs that cover all use cases and all workflow engines.
+* Provide implementation details.
+
+## Design
+
+Core operations (`couler.ops`):
+
+* `run_step(step_def)` where `step_def` is the step definition that can be a container spec, Python function,
```

Comment:
Will this name `step` a little bit confusing since `step` means differently in Tekton and Argo?

In addition, I am a little curious about the design of the DAG in Couler. In Argo, DAG is one of the templates, but in Tekton, there is no concept of templates and users can simply run their tasks in DAG.

Reply:
Great question! Thanks for chiming in.

I think it should be fine as long as we document it properly. In Couler, a `step` simply means a node in the workflow graph (the "smallest" unit in some sense) which is essentially the same for Argo and Tekton. In Tekton, a `task` consists of multiple steps and a `pipeline` consists of multiple `tasks`.

Perhaps we can model the analogy like the following: 
* step (Couler) = step (Argo) = step (Tekton)
* reusable step (Couler) = template (Argo) = task (Tekton)
* workflow (Couler) = workflow (Argo) = pipeline (Tekton)

Even though this may not be 100% accurate but we can try model this as closely as possible to cover most use cases. What do you think?

Reply:
LGTM

Reply:
Just to extend the analogy above to other systems like Airflow, Dagster, and Prefect

| Workflow Engine | Couler        | Argo     | Tekton   | Airflow                                                             | Dagster         | Prefect |
| --------------- | ------------- | -------- | -------- | ------------------------------------------------------------------- | --------------- | ------- |
| Step            | Step          | step     | step     | task                                                                | solid           | task    |
| Composite steps | Reusable step | template | task     | SubDag or [TaskGroup](https://github.com/apache/airflow/pull/10153) | composite solid | ???     |
| Worfklow        | Workflow      | workflow | pipeline | DAG                                                                 | pipeline        | Flow    |

I am not sure about the definition of `Reusable step`, but I assume it means composite steps where it references `task` for Tekton?

I think it will be nice to have a comparison table and keep it updated to help adoption for people from other backends and for implementation for new backends.

Reply:
Thanks! Great idea. I'll add this table to the doc. By "reusable step", I meant parameterized template that can be used to define a step where users only have to specify a few parameters.